File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    public boolean visited[][];
    public boolean visi[][];
    
    public void solve(char[][] board) {
        int m = board.length;
        int n = board[0].length;
        visited = new boolean[m][n];
        visi = new boolean[m][n];
        for (int i = 0;i<m;i++)
        {
            if (board[i][0] == 'O')
            {
                doRoam(board, m, n, i, 0);
            }
            if (board[i][n-1] == 'O')
            {
                doRoam(board, m, n, i, n-1);
            }
        }
        for (int j = 0;j<n;j++)
        {
            if (board[0][j] == 'O')
            {
                doRoam(board, m, n, 0, j);
            }
            if (board[m-1][j] == 'O')
            {
                doRoam(board, m, n, m-1, j);
            }
        }
        for (int i = 0;i<m;i++)
        {
            for (int j = 0;j<n;j++)
            {
                if (board[i][j] == 'O' && visited[i][j] == false)
                {
                    board[i][j] = 'X';
                }
            }
        }
    }
    
    public void doRoam(char[][] board, int m, int n, int i, int j)
    {
        if (i < 0 || j < 0 || i == m || j == n || board[i][j] == 'X')
        {
            return;
        }
        if (visi[i][j] == true || visited[i][j] == true)
        {
            return;
        }
        visited[i][j] = true;
        visi[i][j] = true;
        doRoam(board, m, n, i, j+1);
        doRoam(board, m, n, i+1, j);
        doRoam(board, m, n, i, j-1);
        doRoam(board, m, n, i-1, j);
        visi[i][j] = false;
    }
}

File: 0130-surrounded-regions/0130-surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        rows = len(board)
        cols = len(board[0])
        visited = set()
        
        def dp(i, j):
            if i < 0 or i >= rows or j < 0 or j >= cols or board[i][j] == "X" or (i, j) in visited:
                return
            visited.add((i, j))
            dp(i, j + 1)
            dp(i + 1, j)
            dp(i, j - 1)
            dp(i - 1, j)
            return

        for i in range(rows):
            dp(i, 0)
            dp(i, cols - 1)
        
        for j in range(cols):
            dp(0, j)
            dp(rows - 1, j)

        for i in range(rows):
            for j in range(cols):
                if board[i][j] == "O" and (i, j) not in visited:
                    board[i][j] = "X"
        # rows = len(board)
        # cols = len(board[0])
        # visited = set()
        # conedge = set()
        # def dp(i, j):
        #     if i < 0 or i >= rows or j < 0 or j >= cols or (i, j) in conedge:
        #         return True
        #     if (i, j) in visited or board[i][j] == "X":
        #         return False
        #     visited.add((i, j))
        #     if dp(i, j + 1) or dp(i + 1, j) or dp(i ,j - 1) or dp(i - 1, j):
        #         conedge.add((i, j))
        #         return True
        #     board[i][j] = "X"
        #     return False
        
        # for i in range(rows):
        #     for j in range(cols):
        #         if board[i][j] == "O":
        #             dp(i, j)